[Bug] Netflix H264 not starting properly starting with ISA Omega V 21.5.10

### Describe the problem

Starting with ISA .10 Netflix h.264 Streams doesn't start properly. switching back to .9 is working like expected. Workaround ist to switch the codec to VP9, then there is no problem or switch to Fullscreen/Video and jump 10 Minutes then it is likly to get picture with sound. Language switching sometimes also start the picture. 

Amazon streams doesn't seems to be effected. But due clear reproducing the error, i think it is more an ISA issue then a Netflix-Addon error. 

My System LE 12/ Raspberry Pi 4 

### Possible fix

i don't know, i had also a look into the changes, but don't know to much about wv. 

### Steps to reproduce

Using Omega ISA >= .10
Start netflix stream with h.264. 
System is doing everything like display switch to suitable frequency and spinner is showing. 
Stream is not starting visibly. 
press return, spinner goes away, switch to video/fullscreen
see wild osd, 
jump 10 minutes + 

### Debug log

had no time, maybe not needed because of the small changes, but will deliver one, if necessary

### Stream manifest file(s)

same here

### Additional info

_No response_

### Operating system(s)

Linux

### Operating system version(s)

LE 12.0.2

### InputStream Adaptive version(s)

>= 21.5.10

### Kodi version(s)

Omega

what some busy days. build was successfull, but building addon on 22.3.0 version still not working: 

/opt/libreelec/libreelec.tv/build.LibreELEC-RPi4.aarch64-13.0-devel/build/inputstream.adaptive-22.3.0-Piers/src/codechandler/CodecHandler.cpp:123:22: error: 'class kodi::addon::InputstreamInfo' has no member named 'GetDvccMetadata'
  123 |   if (!(meta == info.GetDvccMetadata()))
      |                      ^~~~~~~~~~~~~~~
/opt/libreelec/libreelec.tv/build.LibreELEC-RPi4.aarch64-13.0-devel/build/inputstream.adaptive-22.3.0-Piers/src/codechandler/CodecHandler.cpp:125:10: error: 'class kodi::addon::InputstreamInfo' has no member named 'SetDvccMetadata'
  125 |     info.SetDvccMetadata(meta);
      |          ^~~~~~~~~~~~~~~

ich not sure if i have time to look further in it today.

as i have not seen some addon updates recently from libreelec addon repo i assume there are some ongoing works behind the scenes

--- a/src/decrypters/widevine/CdmTypeConversion.cpp
+++ b/src/decrypters/widevine/CdmTypeConversion.cpp
@@ -203,37 +203,44 @@ void media::ToCdmInputBuffer(const DEMUX_PACKET* encryptedBuffer,
   inputBuffer->data = encryptedBuffer->pData;
   inputBuffer->data_size = encryptedBuffer->iSize;
   inputBuffer->timestamp = static_cast<int64_t>(encryptedBuffer->pts * 1000000 / STREAM_TIME_BASE);
-  inputBuffer->key_id = encryptedBuffer->cryptoInfo->kid;
-  inputBuffer->key_id_size = 16;
-  inputBuffer->iv = encryptedBuffer->cryptoInfo->iv;
-  inputBuffer->iv_size = 16;
-
-  const uint16_t numSubsamples =
-      encryptedBuffer->cryptoInfo ? encryptedBuffer->cryptoInfo->numSubSamples : 0;
-  if (numSubsamples > 0)
-  {
-    subsamples->reserve(numSubsamples);
-    for (uint16_t i = 0; i < numSubsamples; i++)
-    {
-      subsamples->push_back({encryptedBuffer->cryptoInfo->clearBytes[i],
-                             encryptedBuffer->cryptoInfo->cipherBytes[i]});
-    }
-  }
-
-  inputBuffer->subsamples = subsamples->data();
-  inputBuffer->num_subsamples = numSubsamples;
 
   if (encryptedBuffer->cryptoInfo)
   {
+    DEMUX_CRYPTO_INFO* cryptoInfo = encryptedBuffer->cryptoInfo;
+
+    inputBuffer->key_id = cryptoInfo->kid;
+    inputBuffer->key_id_size = 16;
+    inputBuffer->iv = cryptoInfo->iv;
+    inputBuffer->iv_size = 16;
+
+    const uint16_t numSubsamples = cryptoInfo->numSubSamples;
+    if (numSubsamples > 0)
+    {
+      subsamples->reserve(numSubsamples);
+      for (uint16_t i = 0; i < numSubsamples; i++)
+      {
+        subsamples->push_back({cryptoInfo->clearBytes[i], cryptoInfo->cipherBytes[i]});
+      }
+
+      inputBuffer->num_subsamples = numSubsamples;
+      inputBuffer->subsamples = subsamples->data();
+    }
+    else
+    {
+      inputBuffer->num_subsamples = 0;
+      inputBuffer->subsamples = nullptr;
+    }
+
     inputBuffer->encryption_scheme =
-        ToCdmEncryptionScheme(static_cast<CryptoMode>(encryptedBuffer->cryptoInfo->mode)); // ??
+        ToCdmEncryptionScheme(static_cast<CryptoMode>(cryptoInfo->mode));
+
+    if (inputBuffer->encryption_scheme != cdm::EncryptionScheme::kUnencrypted)
+    {
+      inputBuffer->pattern = {cryptoInfo->cryptBlocks, cryptoInfo->skipBlocks};
+    }
   }
   else
-    inputBuffer->encryption_scheme = cdm::EncryptionScheme::kUnencrypted;
-
-  if (inputBuffer->encryption_scheme != cdm::EncryptionScheme::kUnencrypted)
   {
-    inputBuffer->pattern = {encryptedBuffer->cryptoInfo->cryptBlocks,
-                            encryptedBuffer->cryptoInfo->skipBlocks};
+    inputBuffer->encryption_scheme = cdm::EncryptionScheme::kUnencrypted;
   }
 }

--- a/src/decrypters/widevine/WVCencSingleSampleDecrypter.cpp
+++ b/src/decrypters/widevine/WVCencSingleSampleDecrypter.cpp
@@ -938,6 +938,7 @@ VIDEOCODEC_RETVAL CWVCencSingleSampleDecrypter::DecryptAndDecodeVideo(
   if (sample->cryptoInfo && sample->cryptoInfo->numSubSamples > 0 &&
       (!sample->cryptoInfo->clearBytes || !sample->cryptoInfo->cipherBytes))
   {
+    LOG::LogF(LOGERROR, "Missing crypting clear/cipher bytes info");
     return VC_ERROR;
   }
 

--- a/src/samplereader/FragmentedSampleReader.cpp
+++ b/src/samplereader/FragmentedSampleReader.cpp
@@ -406,7 +406,7 @@ AP4_Result CFragmentedSampleReader::ProcessMoof(AP4_ContainerAtom* moof,
       AP4_ContainerAtom* traf =
           AP4_DYNAMIC_CAST(AP4_ContainerAtom, moof->GetChild(AP4_ATOM_TYPE_TRAF, 0));
 
-      if (!m_protectedDesc || !traf)
+      if (!traf)
         return AP4_ERROR_INVALID_FORMAT;
 
       bool isDefaultProtected{true};

--- a/src/samplereader/FragmentedSampleReader.cpp
+++ b/src/samplereader/FragmentedSampleReader.cpp
@@ -196,8 +196,13 @@ AP4_Result CFragmentedSampleReader::ReadSample()
       // to help convert packets to annex-b
       // it depends on decrypter implementation
       m_sampleData.Reserve(sampleData.GetDataSize());
-      m_singleSampleDecryptor->DecryptSampleData(m_poolId, sampleData, m_sampleData, nullptr, 0,
-                                                 nullptr, nullptr, streamType);
+      if (AP4_FAILED(
+              result = m_singleSampleDecryptor->DecryptSampleData(
+                  m_poolId, sampleData, m_sampleData, nullptr, 0, nullptr, nullptr, streamType)))
+      {
+        Reset(true);
+        return result;
+      }
     }
     else
     {

--- a/src/samplereader/FragmentedSampleReader.cpp
+++ b/src/samplereader/FragmentedSampleReader.cpp
@@ -44,7 +44,7 @@ CFragmentedSampleReader::~CFragmentedSampleReader()
 {
   if (m_singleSampleDecryptor)
     m_singleSampleDecryptor->RemovePool(m_poolId);
-  delete m_decrypter;
+
   delete m_codecHandler;
 }
 
@@ -55,11 +55,10 @@ bool CFragmentedSampleReader::Initialize(SESSION::CStream* stream)
   AP4_SampleDescription* desc{m_track->GetSampleDescription(0)};
   if (desc->GetType() == AP4_SampleDescription::TYPE_PROTECTED)
   {
-    m_protectedDesc = static_cast<AP4_ProtectedSampleDescription*>(desc);
+    auto protectedDesc = static_cast<AP4_ProtectedSampleDescription*>(desc);
 
     AP4_ContainerAtom* schi;
-    if (m_protectedDesc->GetSchemeInfo() &&
-        (schi = m_protectedDesc->GetSchemeInfo()->GetSchiAtom()))
+    if (protectedDesc->GetSchemeInfo() && (schi = protectedDesc->GetSchemeInfo()->GetSchiAtom()))
     {
       AP4_TencAtom* tenc(AP4_DYNAMIC_CAST(AP4_TencAtom, schi->GetChild(AP4_ATOM_TYPE_TENC, 0)));
       if (tenc && tenc->GetDefaultKid())
@@ -133,15 +132,10 @@ AP4_Result CFragmentedSampleReader::ReadSample()
     streamType = DRM::DRMMediaType::AUDIO;
 
   AP4_Result result;
+  AP4_DataBuffer sampleData;
   if (!m_codecHandler->ReadNextSample(m_sample, m_sampleData))
   {
-    bool useDecryptingDecoder =
-        m_protectedDesc &&
-        (m_decrypterCaps.flags & DRM::DecrypterCapabilites::SSD_SECURE_PATH) != 0;
-    bool decrypterPresent{m_decrypter != nullptr};
-    if (AP4_FAILED(result = ReadNextSample(m_track->GetId(), m_sample,
-                                           (m_decrypter || useDecryptingDecoder) ? m_encrypted
-                                                                                 : m_sampleData)))
+    if (AP4_FAILED(result = ReadNextSample(m_track->GetId(), m_sample, sampleData)))
     {
       if (result == AP4_ERROR_EOS)
       {
@@ -170,16 +164,14 @@ AP4_Result CFragmentedSampleReader::ReadSample()
     //AP4_AvcFrameParser::ParseFrameForSPS(m_sampleData.GetData(), m_sampleData.GetDataSize(), 4, sps);
 
     //Protection could have changed in ProcessMoof
-    if (!decrypterPresent && m_decrypter != nullptr && !useDecryptingDecoder)
-      m_encrypted.SetData(m_sampleData.GetData(), m_sampleData.GetDataSize());
-    else if (decrypterPresent && m_decrypter == nullptr && !useDecryptingDecoder)
-      m_sampleData.SetData(m_encrypted.GetData(), m_encrypted.GetDataSize());
+    bool useDecryptingDecoder =
+        m_singleSampleDecryptor &&
+        (m_decrypterCaps.flags & DRM::DecrypterCapabilites::SSD_SECURE_PATH) != 0;
 
     if (m_decrypter)
     {
-      m_sampleData.Reserve(m_encrypted.GetDataSize());
-      if (AP4_FAILED(result =
-                         m_decrypter->DecryptSampleData(m_poolId, m_encrypted, m_sampleData,
+      m_sampleData.Reserve(sampleData.GetDataSize());
+      if (AP4_FAILED(result = m_decrypter->DecryptSampleData(m_poolId, sampleData, m_sampleData,
                                                         NULL, streamType)))
       {
         LOG::Log(LOGERROR, "Decrypt Sample returns failure!");
@@ -200,9 +192,16 @@ AP4_Result CFragmentedSampleReader::ReadSample()
     }
     else if (useDecryptingDecoder)
     {
-      m_sampleData.Reserve(m_encrypted.GetDataSize());
-      m_singleSampleDecryptor->DecryptSampleData(m_poolId, m_encrypted, m_sampleData, nullptr, 0,
+      // Should fall here for the specific case of unencrypted content,
+      // to help convert packets to annex-b
+      // it depends on decrypter implementation
+      m_sampleData.Reserve(sampleData.GetDataSize());
+      m_singleSampleDecryptor->DecryptSampleData(m_poolId, sampleData, m_sampleData, nullptr, 0,
                                                  nullptr, nullptr, streamType);
+    }
+    else
+    {
+      m_sampleData.SetData(sampleData.GetData(), sampleData.GetDataSize());
     }
 
     if (m_codecHandler->Transform(m_sample.GetDts(), m_sample.GetDuration(), m_sampleData,
@@ -235,8 +234,7 @@ uint64_t CFragmentedSampleReader::GetDuration() const
 
 bool CFragmentedSampleReader::IsEncrypted() const
 {
-  return (m_decrypterCaps.flags & DRM::DecrypterCapabilites::SSD_SECURE_PATH) != 0 &&
-         m_decrypter != nullptr;
+  return (m_decrypterCaps.flags & DRM::DecrypterCapabilites::SSD_SECURE_PATH) != 0 && m_decrypter;
 }
 
 bool CFragmentedSampleReader::GetInformation(kodi::addon::InputstreamInfo& info)
@@ -405,9 +403,6 @@ AP4_Result CFragmentedSampleReader::ProcessMoof(AP4_ContainerAtom* moof,
       AP4_CencSampleInfoTable* sample_table{nullptr};
       AP4_UI32 algorithm_id = 0;
 
-      delete m_decrypter;
-      m_decrypter = 0;
-
       AP4_ContainerAtom* traf =
           AP4_DYNAMIC_CAST(AP4_ContainerAtom, moof->GetChild(AP4_ATOM_TYPE_TRAF, 0));
 
@@ -446,7 +441,7 @@ AP4_Result CFragmentedSampleReader::ProcessMoof(AP4_ContainerAtom* moof,
       if (!m_singleSampleDecryptor)
         return AP4_ERROR_INVALID_PARAMETERS;
 
-      m_decrypter = new CAdaptiveCencSampleDecrypter(m_singleSampleDecryptor, sample_table);
+      m_decrypter = std::make_unique<CAdaptiveCencSampleDecrypter>(m_singleSampleDecryptor, sample_table);
 
       // Inform decrypter of pattern decryption (CBCS)
       AP4_UI32 schemeType = m_protectedDesc->GetSchemeType();
@@ -472,6 +467,7 @@ AP4_Result CFragmentedSampleReader::ProcessMoof(AP4_ContainerAtom* moof,
     else
     {
       // Reset for unencrypted content
+      m_decrypter.reset();
       m_readerCryptoInfo = CryptoInfo();
     }
   }

--- a/src/samplereader/FragmentedSampleReader.cpp
+++ b/src/samplereader/FragmentedSampleReader.cpp
@@ -195,6 +195,8 @@ AP4_Result CFragmentedSampleReader::ReadSample()
       // Should fall here for the specific case of unencrypted content,
       // to help convert packets to annex-b
       // it depends on decrypter implementation
+      //! @todo: it appears to have been implemented in the past for Widevine
+      //!        and may not be suitable for other decryptors
       m_sampleData.Reserve(sampleData.GetDataSize());
       if (AP4_FAILED(
               result = m_singleSampleDecryptor->DecryptSampleData(

--- a/src/samplereader/FragmentedSampleReader.h
+++ b/src/samplereader/FragmentedSampleReader.h
@@ -70,12 +70,11 @@ private:
   uint64_t m_timeBaseExt{0};
   uint64_t m_timeBaseInt{0};
   AP4_Sample m_sample;
-  AP4_DataBuffer m_encrypted;
   AP4_DataBuffer m_sampleData;
   CodecHandler* m_codecHandler{nullptr};
   std::vector<uint8_t> m_defaultKey;
   AP4_ProtectedSampleDescription* m_protectedDesc{nullptr};
   std::shared_ptr<Adaptive_CencSingleSampleDecrypter> m_singleSampleDecryptor{nullptr};
-  CAdaptiveCencSampleDecrypter* m_decrypter{nullptr};
+  std::unique_ptr<CAdaptiveCencSampleDecrypter> m_decrypter;
   CryptoInfo m_readerCryptoInfo{};
 };

